lib/matplotlib/cm.py

```diff
@@ -255,10 +254,18 @@ def to_rgba(self, x, alpha=None, bytes=False, norm=True):
                     xx = x
                 else:
                     raise ValueError("third dimension must be 3 or 4")
-                if bytes and xx.dtype != np.uint8:
-                    xx = (xx * 255).astype(np.uint8)
-                if not bytes and xx.dtype == np.uint8:
-                    xx = xx.astype(float) / 255
+                if xx.dtype.kind == 'f':
+                    if xx.max() > 1 or xx.min() < 0:
```

Comment:
I think uses like scatter are irrelevant because the checking in question is invoked only for MxNx3 and MxNx4 inputs, which are assumed to be RGB or RGBA images.
As for "impossible": `to_rgba` is not private, so it is entirely possible for a user to generate a 3-D float array that would fail the check.  If such an array is fed to `imshow` with interpolation 'none', will it be run through `to_rgba`?  Certainly it will not be normalized.  
The options are to remove the check on the grounds that it is usually unnecessary (we have managed without it for a long time), to leave it in and accept the performance penalty, or to find a way to distinguish between cases where it might be useful, if such cases exist, and cases where we are sure it is pointless.  The last probably would be the best *if* it can be accomplished with minimal added complexity.